### ISSUE TITLE
Emoji domain not recognised as link 

### ISSUE THREAD

**Issue by christopheschwyzer** [signalapp/Signal-Android#11247]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Links that contain emojis are not recognised as proper links.

### Steps to reproduce
- Open Signal
- Go to any chat conversation or *Note to Self*
- Share a link that contains an emoji. E.g. https://🤫.st 

**Actual result:** 
There is no link preview, the URL is not recognised as a proper link (not clickable)

**Expected result:** Valid domains should be handled as proper links. 

It's a bit edge case bug, and might even count as feature request. However I'd be happy for a fix - which most likely goes here somewhere: https://github.com/signalapp/Signal-Android/blob/dc6dc192dc4c846586152b2b8949aaebf8be5244/app/src/main/java/org/thoughtcrime/securesms/linkpreview/LinkPreviewUtil.java#L35 cc @greyson-signal @alan-signal Unfortunately I'm not familiar with Java, otherwise I'd do a PR myself. Let me know if I can help nonetheless. 

 

### Screenshots
![Screenshot_20210429-224722](https://user-images.githubusercontent.com/1888817/116617317-fdbc3000-a93d-11eb-81fe-7e00b16945d2.png)

### Device info
**Device:** OnePlus 6
**Android version:** 10
**Signal version:** 5.7.1

### Link to debug log
Nothing to log

**Comment by hiqua** [comment]
With your URL this variable is `false`:
```
      boolean validCharacters = ALL_ASCII_PATTERN.matcher(cleanedDomain).matches() ||
                                ALL_NON_ASCII_PATTERN.matcher(cleanedDomain).matches();
```
because `cleanedDomain` ends up being `🤫st`. Not sure what's the reasoning behind this check though...

**Comment by mayjs** [comment]
@hiqua I noticed a similar issue with domains containing umlauts (e.g. https://üei.de). Could this be related?

**Comment by hiqua** [comment]
I would say so.

**Comment by stophecom** [comment]
Yes, it is this still relevant. Links with emoji or umlauts (ä, ö, ü) are not recognized as proper links.

**Comment by ckujau** [comment]
Jup, it's still relevant. Btw, this affects the desktop version as well.
![Screenshot from 2022-03-27 19-55-10](https://user-images.githubusercontent.com/822519/160294201-d09a4094-f2bf-468f-ba42-763ca1afb60e.png)

**Comment by greyson-signal** [comment]
We use the stock android linkifier. I don't anticipate that we'll be making a custom one anytime soon, apologies.

**Comment by ckujau** [comment]
Hm, both links work in Google Chat (formerly "Hangouts") just fine. And also in other applicable (e.g. "Slack")
![Screenshot_20220328-232524~2](https://user-images.githubusercontent.com/822519/160490261-eeeca8bf-4ce0-4a7a-88e8-0e25e838adbb.png)

**Comment by greyson-signal** [comment]
Ah, I'm sorry, I misspoke. So we do use the default linkifier, _but_ we filter out some results that fail certain rules. Right now, the main rule is that we don't linkify links whose domain has a mix of ascii and non-ascii characters. This is to help prevent [homograph attacks](https://en.wikipedia.org/wiki/IDN_homograph_attack). Our rule for that is a little coarse, but honestly given the huge character space we're trying to play it safe here.

So summary:
- If a link is linkified, it's because the default linkifier identified it as a link
- If it's _not_ linkified, it's because the default linkifier missed it _or_ it violates one of the rules we setup to prevent homograph attacks.

That said, I still think the outcome for this ticket is the same: I don't anticipate we'll do more nuanced stuff here just to be safe. But apologies for the previous incorrect reasoning.

**Comment by stophecom** [comment]
Thanks for the explanation. 
I would argue it's a nice trick and this rule definitely helps. But it's also a bit arbitrary. And with newer TLDs this "trick" might become less effective. E.g. 

<img width="139" alt="image" src="https://user-images.githubusercontent.com/1888817/160789031-ae5c5cb9-13de-42bf-9781-2540fae35788.png">

And what about my emoji domain 😭😂😂😂. So sad.
